### ISSUE TITLE
==> Error: develop does not satisfy 1.9.1.0

### ISSUE THREAD

**Issue by jgalarowicz** [spack/spack#7913]
At the request of the larger OpenSpeedShop/CBTF team, I'm trying to create a path through the Krell tools that is more version specific for a release (1.9.1.0).   When I add specific version information based on a branch name, I'm getting this error:  
```
==> Error: develop does not satisfy 1.9.1.0
```
Here is the cbtf/package.py file.  I have four other package files that get the same message, so I must be doing something fundamentally wrong.   Can anyone see an issue?    Thanks.

```
##############################################################################
# Copyright (c) 2013-2018, Lawrence Livermore National Security, LLC.
# Produced at the Lawrence Livermore National Laboratory.
#
# This file is part of Spack.
# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
# LLNL-CODE-647188
#
# For details, see https://github.com/spack/spack
# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License (as
# published by the Free Software Foundation) version 2.1, February 1999.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
# conditions of the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
##############################################################################
##########################################################################
# Copyright (c) 2015-2018 Krell Institute. All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA
##########################################################################

from spack import *


class Cbtf(CMakePackage):
    """CBTF project contains the base code for CBTF that supports creating
       components, component networks and the support to connect these
       components and component networks into sequential and distributed
       network tools.

    """
    homepage = "http://sourceforge.net/p/cbtf/wiki/Home"

    # Use when the git repository is available

    version('1.9.1.0', branch='1.9.1.0',
            git='https://github.com/OpenSpeedShop/cbtf.git')

    version('1.9.1', branch='master',
            git='https://github.com/OpenSpeedShop/cbtf.git')

    version('develop', branch='master',
            git='https://github.com/OpenSpeedShop/cbtf.git')

    variant('cti', default=False,
            description="Build MRNet with the CTI startup option")

    variant('runtime', default=False,
            description="build only the runtime libraries and collectors.")

    variant('build_type', default='None', values=('None'),
            description='CMake build type')

    depends_on("cmake@3.0.2:", type='build')

    depends_on("boost@1.66.0", when='@1.9.1.0')
    depends_on("boost@1.50.0:", when='@develop')

    # For MRNet
    depends_on("mrnet@5.0.1-3:+cti", when='@develop+cti')
    depends_on("mrnet@5.0.1-3:+lwthreads", when='@develop')
    depends_on("mrnet@5.0.1-3:+cti", when='@1.9.1.0+cti')
    depends_on("mrnet@5.0.1-3:+lwthreads", when='@1.9.1.0')

    # For Xerces-C
    depends_on("xerces-c@3.1.1:", when='@develop')
    depends_on("xerces-c@3.1.4", when='@1.9.1.0')

    # Work around for spack libxml2 package bug, take off python when fixed
    depends_on("libxml2+python", when='@develop')
    depends_on("libxml2@2.9.4+python", when='@1.9.1.0')

    parallel = False

    build_directory = 'build_cbtf'

    def cmake_args(self):

        spec = self.spec

        # Boost_NO_SYSTEM_PATHS  Set to TRUE to suppress searching
        # in system paths (or other locations outside of BOOST_ROOT
        # or BOOST_INCLUDEDIR).  Useful when specifying BOOST_ROOT.
        # Defaults to OFF.

        compile_flags = "-O2 -g"

        if spec.satisfies('+runtime'):

            # Install message tag include file for use in Intel MIC
            # cbtf-krell build
            # FIXME
            cmake_args = [
                '-DCMAKE_CXX_FLAGS=%s'     % compile_flags,
                '-DCMAKE_C_FLAGS=%s'       % compile_flags,
                '-DRUNTIME_ONLY=TRUE',
                '-DBoost_NO_SYSTEM_PATHS=TRUE',
                '-DXERCESC_DIR=%s'         % spec['xerces-c'].prefix,
                '-DBOOST_ROOT=%s'          % spec['boost'].prefix,
                '-DMRNET_DIR=%s'           % spec['mrnet'].prefix,
                '-DCMAKE_MODULE_PATH=%s'   % join_path(
                    prefix.share, 'KrellInstitute', 'cmake')]
        else:
            cmake_args = [
                '-DCMAKE_CXX_FLAGS=%s'     % compile_flags,
                '-DCMAKE_C_FLAGS=%s'       % compile_flags,
                '-DBoost_NO_SYSTEM_PATHS=TRUE',
                '-DXERCESC_DIR=%s'         % spec['xerces-c'].prefix,
                '-DBOOST_ROOT=%s'          % spec['boost'].prefix,
                '-DMRNET_DIR=%s'           % spec['mrnet'].prefix,
                '-DCMAKE_MODULE_PATH=%s'   % join_path(
                    prefix.share, 'KrellInstitute', 'cmake')]

        return cmake_args
```

**Comment by davydden** [comment]
> develop does not satisfy 1.9.1.0

can you try to simplify this file to just a few lines which illustrate the problem? I don't immediately see what's wrong.

What I can tell, though, is that `@1.9.1.0:` will certainly be satisfied by `develop` as it's larger than any numeric versions. There are unit tests in Spack which make sure this is not broken and we also use it in a number of packages, for example [here](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dealii/package.py#L118) (note that there is no `9.0` version, so at the moment this can only be satisfied by the `develop`).

**Comment by jgalarowicz** [comment]
@davydden Thank you for the reply.  I will try to shrink this.   So, what you are saying is that if I don't have a 1.9.1.0 for every package, spack will try to "satisfy" the dependency by using another version?

**Comment by davydden** [comment]
> if I don't have a 1.9.1.0 for every package, spack will try to "satisfy" the dependency by using another version?

I am not sure I follow, what I meant is that if you have a requirements somewhere like
```
depends_on(boost, when='@1.9.1.0:`)
```
this means that anything from 1.9.1.0 and up will satisfy this. So does `develop`.

**Comment by jgalarowicz** [comment]
@davydden @adamjstewart  I found the problem, I had an incorrect depends_on clause where 1.9.1.0 was dependent on develop.
```+    depends_on("cbtf@develop+cti", when='@1.9.1.0+cti')
```
Once, I straightened that out, I can build again.
I was thrown off because the error was in the cbtf-argonavis package file not the cbtf package which I was trying to build.
Sorry for the bother, but your comments helped me figure it out.
Thanks again!
Jim G
